### ISSUE TITLE
Qubes builder fails with sign-compare error when trying to build fc30 template

### ISSUE THREAD

**Issue by iprid23** [QubesOS/qubes-issues#5217]
**Qubes OS version**
R4.0.1

**Affected component(s) or functionality**
qubes-builder
vmm-xen-vm-fc-30
core-qubesdb-vm-fc30

**Brief summary**
I had to manually change my builder.conf to enable fc30 instead of fc29. When building I noticed that -Wsign-compare CFLAG is set together with -Werror which means that a sign-compare warning is counted as an error leading to a failed build with certain packages.

**To Reproduce**
1. Try ```make  vmm-xen-vm-fc30``` or ```make core-qubesdb-vm-fc30```

**Expected behavior**
The build should not fail

**Actual behavior**
The build failed with the following messages:
vmm-xen-vm-fc30:
````gcc -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -m64 -DBUILD_ID -fno-strict-aliasing -std=gnu99 -Wall -Wstrict-prototypes -Wdeclaration-after-statement -Wno-unused-but-set-variable -Wno-unused-local-typedefs -O2 -fomit-frame-pointer -D__XEN_INTERFACE_VERSION__=__XEN_LATEST_INTERFACE_VERSION__ -MMD -MF .build.d -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -O1 -fPIC -I../../tools/include -I../../tools/libs/toollog/include -I../../tools/libs/evtchn/include -I../../tools/libxc/include -Ixen/lowlevel/xc -I/usr/include/python3.7m -c xen/lowlevel/xc/xc.c -o build/temp.linux-x86_64-3.7/xen/lowlevel/xc/xc.o -fno-strict-aliasing -Werror
xen/lowlevel/xc/xc.c: In function ‘pyxc_domain_create’:
xen/lowlevel/xc/xc.c:138:24: error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
  138 |         for ( i = 0; i < sizeof(xen_domain_handle_t); i++ )
      |                        ^
xen/lowlevel/xc/xc.c: In function ‘pyxc_domain_sethandle’:
xen/lowlevel/xc/xc.c:295:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
  295 |     for ( i = 0; i < sizeof(xen_domain_handle_t); i++ )
      |                    ^
xen/lowlevel/xc/xc.c: In function ‘pyxc_domain_getinfo’:
xen/lowlevel/xc/xc.c:374:24: error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
  374 |         for ( j = 0; j < sizeof(xen_domain_handle_t); j++ )
      |                        ^
xen/lowlevel/xc/xc.c: In function ‘pyxc_get_device_group’:
xen/lowlevel/xc/xc.c:660:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘uint32_t’ {aka ‘unsigned int’} [-Werror=sign-compare]
  660 |     for ( i = 0; i < num_sdevs; i++ )
      |                    ^
xen/lowlevel/xc/xc.c: In function ‘pyxc_physinfo’:
xen/lowlevel/xc/xc.c:988:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
  988 |     for ( i = 0; i < sizeof(pinfo.hw_cap)/4; i++ )
      |                    ^
cc1: all warnings being treated as errors
error: command 'gcc' failed with exit status 1
make: *** [Makefile:12: build] Error 1
````
core-qubesdb-vm-fc30:
````
cc -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -I../include -I/usr/include/python3.7m -c qubesdb.c -o build/temp.linux-x86_64-3.7/qubesdb.o -fno-strict-aliasing -Werror
qubesdb.c: In function ‘qdbpy_list’:
qubesdb.c:174:23: error: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Werror=sign-compare]
  174 |         for (i = 0; i < list_len; i++) {
      |                       ^
qubesdb.c: In function ‘qdbpy_multiread’:
qubesdb.c:217:23: error: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Werror=sign-compare]
  217 |         for (i = 0; i < list_len; i++) {
      |                       ^
qubesdb.c: In function ‘parse_handle_path’:
qubesdb.c:412:23: error: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Werror=sign-compare]
  412 |     if (strlen(*path) != path_len) {
      |                       ^~
cc1: all warnings being treated as errors
error: command 'cc' failed with exit status 1
make: *** [Makefile:8: build] Error 1
make: Leaving directory '/home/user/rpmbuild/BUILD/qubes-db-4.0.10/python'
````
**Solutions you've tried**
For core-qubesdb-vm-fc30 I went and changed the ```i``` to unsigned int in the qubes-src folder and that fixed it. I tried doing the same for the vmm-xen but since the source comes from an archive it's overwritten each time rpmbuild does its job so I am thinking of making a patch and adding it to the spec or just finding a way to disable -Werror.

**Comment by iprid23** [comment]
Trying to add -Wno-error=sign-compare to xen.spec.in now.

**Comment by iprid23** [comment]
The setup.py patching with -Wno-error=sign-compare solved the issue and vmm-xen-vm-fc30 was built successfully. Now there's a new problem when doing ```make template```

````
Error: 
 Problem 1: package qubes-libvchan-xen-4.0.5-1.fc30.x86_64 requires libxenctrl.so.4.11()(64bit), but none of the providers can be installed
  - package qubes-libvchan-xen-4.0.5-1.fc30.x86_64 requires libxenvchan.so.4.11()(64bit), but none of the providers can be installed
  - cannot install both xen-libs-4.11.2-1.fc30.x86_64 and xen-libs-2001:4.8.5-7.fc30.x86_64
  - cannot install both xen-libs-4.11.1-4.fc30.x86_64 and xen-libs-2001:4.8.5-7.fc30.x86_64
  - package xen-qubes-vm-2001:4.8.5-7.fc30.x86_64 requires libxenctrl.so.4.8()(64bit), but none of the providers can be installed
  - package xen-qubes-vm-2001:4.8.5-7.fc30.x86_64 requires libxenguest.so.4.8()(64bit), but none of the providers can be installed
  - package xen-qubes-vm-2001:4.8.5-7.fc30.x86_64 requires libxenlight.so.4.8.1()(64bit), but none of the providers can be installed
  - package xen-qubes-vm-2001:4.8.5-7.fc30.x86_64 requires libxlutil.so.4.8()(64bit), but none of the providers can be installed
  - package xen-qubes-vm-2001:4.8.5-7.fc30.x86_64 requires xen-libs = 2001:4.8.5-7.fc30, but none of the providers can be installed
  - package qubes-core-agent-4.0.47-1.fc30.x86_64 requires qubes-libvchan, but none of the providers can be installed
  - package qubes-vm-dependencies-4.0.5-1.fc30.noarch requires xen-qubes-vm, but none of the providers can be installed
  - package qubes-vm-dependencies-4.0.5-1.fc30.noarch requires qubes-core-agent, but none of the providers can be installed
  - conflicting requests
 Problem 2: package qubes-libvchan-xen-4.0.5-1.fc30.x86_64 requires libxenctrl.so.4.11()(64bit), but none of the providers can be installed
  - package qubes-libvchan-xen-4.0.5-1.fc30.x86_64 requires libxenvchan.so.4.11()(64bit), but none of the providers can be installed
  - cannot install both xen-libs-4.11.2-1.fc30.x86_64 and xen-libs-2001:4.8.5-7.fc30.x86_64
  - cannot install both xen-libs-4.11.1-4.fc30.x86_64 and xen-libs-2001:4.8.5-7.fc30.x86_64
  - package qubes-db-4.0.10-1.fc30.x86_64 requires xen-libs >= 2001:4.6.1-20, but none of the providers can be installed
  - package qubes-db-vm-4.0.10-1.fc30.x86_64 requires qubes-db, but none of the providers can be installed
  - package qubes-core-agent-4.0.47-1.fc30.x86_64 requires qubes-db-vm, but none of the providers can be installed
  - package qubes-core-agent-dom0-updates-4.0.47-1.fc30.x86_64 requires qubes-core-agent = 4.0.47, but none of the providers can be installed
  - package pulseaudio-qubes-4.0.24-1.fc30.x86_64 requires libvchan-xen.so()(64bit), but none of the providers can be installed
  - package qubes-vm-recommended-4.0.5-1.fc30.noarch requires qubes-core-agent-dom0-updates, but none of the providers can be installed
  - package qubes-vm-recommended-4.0.5-1.fc30.noarch requires pulseaudio-qubes, but none of the providers can be installed
  - conflicting requests
````

**Comment by marmarek** [comment]
> qubes-libvchan-xen-4.0.5-1.fc30.x86_64 requires libxenctrl.so.4.11()(64bit)

You've build core-vchan-xen with wrong xen version (probably you've done that before building vmm-xen). Simply rebuild core-vchan-xen.

**Comment by iprid23** [comment]
@marmarek thanks for the suggestion, I did rebuild ```core-libvirt core-vchan-xen core-qubesdb``` after building vmm-xen, but I still get the error when I try to do make template:
```
 Problem 1: package qubes-libvchan-xen-4.0.5-1.fc30.x86_64 requires libxenctrl.so.4.11()(64bit), but none of the providers can be installed
```

Unrelated but again part of the builder: In the meantime I also tried to build through sys-whonix gateway but cannot recv the gpg keys:
```
gpg --keyserver pgp.mit.edu --recv-keys 916B8D99C38EAF5E8ADC7A2A8D66066A2EEACCDA
gpg: keyserver receive failed: Server indicated a failure
```
Trying with other servers does not help.

**Comment by marmarek** [comment]
Hmm, maybe you have still xen-devel-4.11 in chroot? Try removing it before build, like this: `sudo chroot chroot-vm-fc30 dnf remove xen-devel`. Or remove the whole chroot with `make clean-chroot`

**Comment by iprid23** [comment]
@marmarek thanks, actually a faster idea was to just touch some file that makes the chroot to be rebuilt again but faster. You wrote that in some some other issue. Anyway indeed there was xen-devel-4.11, deleting it solved the issue.

1. However I can confirm that both **core-qubesdb** and **xen-vmm** need the **-Wno-error=sign-compare** patches to compile for fc30.
2. There are other issues, I've noticed:
- whonix builder repo from QubesOS Github organization is private, so you have to put in the path to the Whonix github repo in builder.conf and use that repo instead:
https://github.com/Whonix/qubes-template-whonix
- **grubby-dummy** must be build before **linux-template-builder** despite being listed after it in the component list
- there is no explanation as to how the .iso is signed and no sign file for the iso either.
**I assume the template is being built before the make sign-all, so we do not sign the packages inside the template rpm, but we sign the template rpm?**
- there is no explanation on how to use SplitGPG for rpm sign, are there any special steps?
- when building qubes-vm, how do I select which template to build?

If you want, I can open some of those as additional issues here.

Regards

**Comment by marmarek** [comment]
>     * whonix builder repo from QubesOS Github organization is private

It isn't private. It doesn't exist. The canonical place for it is in Whonix organization. There is also fork on my account (it's set by default in example-configs/templates.conf).

>     * **grubby-dummy** must be build before **linux-template-builder** despite being listed after it in the component list

Fixed here: https://github.com/QubesOS/qubes-builder/commit/36307f761ea21e330c824e0908c552640c852f31

>     * there is no explanation as to how the .iso is signed and no sign file for the iso either.
>       **I assume the template is being built before the make sign-all, so we do not sign the packages inside the template rpm, but we sign the template rpm?**

If you build it in one go, then yes. In practice signing packages inside is irrelevant, as the template doesn't keep individual rpm packages itself, nor an information about its signatures.

> there is no explanation on how to use SplitGPG for rpm sign, are there any special steps?

See here: https://github.com/QubesOS/qubes-builder/blob/master/doc/SplitGPG.md

> when building qubes-vm, how do I select which template to build?

With DISTS_VM variable. You can either set it in builder.conf, or provide in environment variable (in that case make sure you don't use `=` to set it in builder.conf, `?=` is fine - the default).

**Comment by marmarek** [comment]
>     1. However I can confirm that both **core-qubesdb** and **xen-vmm** need the **-Wno-error=sign-compare** patches to compile for fc30.

I can confirm it too. I definitely don't want disabling that warning, rather fix it properly instead.

**Comment by iprid23** [comment]
Agreed. I looked at the code and it was fairly easy for core-qubesdb(just change the iterator integers to unsigned, since they are always positive), while it toke more time to fix it for vmm-xen. Thanks for the other clarifications.

As for whonix, by default it tried to pull from https://github.com/QubesOS/qubes-template-whonix by default and I had to override it in builder.conf.

Anyway, some more questions:
Can I have the qubes-builder in some other dir and not ~/qubes-builder?
If I build a template with one qubes-builder instance, how can I put it in another qubes-builder instance, so I can get an ISO?
Say I have ~/qubes-builder-dom0(only dom0) and ~/qubes-builder-whonix(only whonix).
Can I get the template RPM from ~/qubes-builder-whonix and put in some dir in ~/qubes-builder-dom0, so I can ask ~/qubes-builder-dom0 to build the ISO?

And one last thing, are qubes packages(vmm-xen, core-qubesdb, etc.) reproducible or are there any plans to make them reproducible?
Regards
